eeps/eep-0076.md

```diff
@@ -0,0 +1,118 @@
+    Author: Richard Carlsson <carlsson.richard(at)gmail(dot)com>
+    Status: Draft
+    Type: Standards Track
+    Created: 10-Jan-2025
+    Erlang-Version: OTP-28.0
+    Post-History:
+****
+EEP 76: Assignment in Comprehensions
+----
+
+Abstract
+========
+
+This EEP adds assignments to comprehension qualifier lists, providing
+a convenient and readable alternative to other syntactical tricks.
+
+Rationale
+=========
+
+It would often be useful to be able to easily bind variables in the
+qualifier sequence of a comprehension, for example:
+
+```erlang
+[Z || X <- Ls,
+      {foo, Y} = g(X),
+      Z <- f(Y, h(Y))].
+```
+
+using plain `Pattern = ...,` entries between qualifiers.
+
+You can do this today by writing a singleton generator:
+
+```erlang
+[Z || X <- Ls,
+      {foo, Y} <- [g(X)],  % produce single element
+      Z <- f(Y, h(Y))].
+```
+
+but this has some drawbacks:
+
+- The intent is not clear to the reader
+- You have to remember to write a single element list around the
+  right hand side argument (which itself could be a list)
+- You probably want it to be a strict generator so typos don't just
+  silently yield no elements
+- Someone editing the code later may accidentally add extra
+  elements to the right hand side list, causing unintended Cartesian
+  combinations
+
+Another trick is to piggy-back on a boolean test, using export of
+bindings from within a subexpression to propagate to the subsequent
+qualifiers:
+
+```erlang
+[Z || X <- Ls,
+      is_tuple(Y = g(X)),
+      Z <- f(Y, h(element(2,Y)))].
+```
+
+which is very much not a recommended style in Erlang, making it hard
+to see where the bindings come from. It also relies on having a
+suitable test as part of the qualifiers for the value you want to
+bind. If you don't, it is possible to invent a dummy always-true test:
+
+```erlang
+[Z || X <- Ls,
+      foobar =/= (Y = g(X)),
+      Z <- f(Y, h(Y))].
+```
+
+Such tricks are very bad for readability and maintenance of the code,
+making the logic hard to follow.
+
+Specification
+========================
+
+It is in fact already allowed syntactically to have a `Pattern = ...`
+match expression in the qualifiers. This however gets interpreted as
+any other expression - thus expected to produce a boolean value - and
+if `false`, the current element will be skipped.
+
+Hence, any qualifier following after a match `Var = Expr` will only be
+executed if `Var` has the value `true`. We can therefore expect that
+no such uses exist in practice, because `Var` would be fully
+redundant. (The OTP code base has been checked and does not contain
+any.)
```

Comment:
I'd say this is still a backwards incompatible change, because if someone relies on it, the behaviour will change in very subtle ways:

```
2> [{X, Y} || X <- [1, 2.0, 3], Y = is_integer(X)].
[{1,true},{3,true}]
```

Also, for completeness, one could introduce maybe expressions' `?=`operator in comprehensions, which would work similar to an assignment + filter (as it will discard the element if it does not match the pattern).

Reply:
It is an incompatible change, yes. To protect against that, one could start with making `Pat = ...` an error in OTP 28, and then add the new interpretation in OTP 29. I really did not want to add yet another special token. (`?=` could be useful though). One possibility is to use the so far reserved token 'let' as a prefix: `let Pat = ..., ...`.

Reply:
There's an option of making it an error in 28, and introducing an opt-in feature that uses the new semantics (rather than erroring), and in 29 making this feature the default.
This way, it won't silently give errors, but won't delay the new syntax for another release for people that are ready for it.

Reply:
Sounds like a winner to me.